types/src/keyless/proof_simulation.rs

```diff
@@ -0,0 +1,259 @@
+    let g2 = g2_projective_str_to_affine(["7060239192912576352445678919251015303857900508169996987700616563495505759758", "2459845072558806286978423063428307489778927966556743480120663459709217599487"], ["19288633317757364243662951827532421887714035432540311650844990893553936393814", "20639282316004454458884347800936381746504150536012576786666607919028441606072"]).unwrap();
+    let alpha = ark_ff::Fp::<MontBackend<FrConfig, 4>, 4>::from_bigint(alpha).unwrap();
+    let beta = ark_ff::Fp::<MontBackend<FrConfig, 4>, 4>::from_bigint(beta).unwrap();
+    let delta = ark_ff::Fp::<MontBackend<FrConfig, 4>, 4>::from_bigint(delta).unwrap();
+    let gamma = ark_ff::Fp::<MontBackend<FrConfig, 4>, 4>::from_bigint(gamma).unwrap();
+    let pk =  ProvingKeyWithTrapdoor { 
+        gamma_abc_g1: gamma_abc_g1.clone(),
+        alpha,
+        beta,
+        delta,
+        gamma,
+        g1,
+        g2
+    };
+    let alpha_g1 = g1_projective_str_to_affine("5572059596569521478142909013551365241483584539326713643538402534559131771215", "17730641409534717358676668589645204443673285030614991773453774266753084779839").unwrap();
+    let beta_g2 = g2_projective_str_to_affine(["18770618686917993373652785848897272442830690230800448000834753889342693280548", "1244082553567860317529082195476871724475625917678564162594525799206784796895"], ["2130688070722815857544427076253407755416575070301508652936230484102777632154", "6104769200283876349074000611313817829507631277250251203644007998595148003804"]).unwrap();
+    let gamma_g2 = g2_projective_str_to_affine(["13321756384019475282834053010962858734065256385792198252178574019857707055625", "9904540203481000972785329888895853465145640470161185325980745361477345980499"],["13183258375250648244090549119792217999633468590401818473812106012080096645793", "9163822098487266592309953971558453292100379671136954613307467823219261972973"]).unwrap();
+    let delta_g2 = g2_projective_str_to_affine(["9263958447477535187142724208180520744776704295633711436406632372106465499165", "6807912405557884826193725256367335580321369623359346147279599622449143736970"], ["18838367547891272887641438914091432084648683803724358191808362736715304958346", "7114529694217827778623886772036286266862226319773425773369673734499262479817"]).unwrap();
+    let vk = VerifyingKey { alpha_g1, beta_g2, gamma_g2, delta_g2, gamma_abc_g1 };
+    let pvk = prepare_verifying_key::<E>(&vk);
+
+    let mut rng = ark_std::rand::rngs::StdRng::seed_from_u64(test_rng().next_u64());
+    for _ in 0..n_iters {
+        let proof = Groth16Simulator::<E>::create_random_proof_with_trapdoor(
+            &[public_input],
+            &pk,
+            &mut rng,
+        )
+        .unwrap();
+
+        assert!(Groth16::<E>::verify_with_processed_vk(&pvk, &[public_input], &proof).unwrap());
+        let a = E::ScalarField::rand(&mut rng);
+        assert!(!Groth16::<E>::verify_with_processed_vk(&pvk, &[a], &proof).unwrap());
+    }
+}
```

Comment:
Can you test `circuit_agnostic_setup_with_trapdoor` too: generate a VK, generate a proof for `x`, make sure it verifies for `x`, make sure it does not verify for `x'`?